packages/dataviews/src/view-grid.js

```diff
@@ -7,21 +7,123 @@ import classnames from 'classnames';
  * WordPress dependencies
  */
 import {
-	__experimentalGrid as Grid,
 	__experimentalHStack as HStack,
 	__experimentalVStack as VStack,
+	privateApis as componentsPrivateApis,
 	Tooltip,
 } from '@wordpress/components';
-import { __ } from '@wordpress/i18n';
-import { useAsyncList } from '@wordpress/compose';
-import { useState } from '@wordpress/element';
+import {
+	useAsyncList,
+	useInstanceId,
+	useResizeObserver,
+} from '@wordpress/compose';
+import {
+	createContext,
+	useContext,
+	useEffect,
+	useMemo,
+	useRef,
+	useState,
+} from '@wordpress/element';
+import { __, isRTL } from '@wordpress/i18n';
+import { isAppleOS } from '@wordpress/keycodes';
 
 /**
  * Internal dependencies
  */
+import { unlock } from './lock-unlock';
 import ItemActions from './item-actions';
 import SingleSelectionCheckbox from './single-selection-checkbox';
 
+const {
+	useCompositeStoreV2: useCompositeStore,
+	CompositeV2: Composite,
+	CompositeItemV2: CompositeItem,
+	CompositeRowV2: CompositeRow,
+} = unlock( componentsPrivateApis );
+
+const GridContext = createContext( {} );
+
+function Grid( { id, children, ...gridProps } ) {
+	const baseId = useInstanceId( Grid, 'view-grid', id );
+	const gridRef = useRef( null );
+	const store = useCompositeStore( {
+		focusWrap: 'horizontal',
+		rtl: isRTL(),
+	} );
+	const context = useMemo(
+		() => ( { ...store, baseId } ),
+		[ store, baseId ]
+	);
+
+	return (
+		<Composite
+			id={ baseId }
+			ref={ gridRef }
+			role="grid"
+			store={ store }
+			{ ...gridProps }
+		>
+			<GridContext.Provider value={ context }>
+				<GridRows baseId={ baseId } gridRef={ gridRef }>
+					{ children }
+				</GridRows>
+			</GridContext.Provider>
+		</Composite>
+	);
+}
+
+function GridRows( { baseId, gridRef, children } ) {
+	const [ columnCount, setColumnCount ] = useState( children?.length || 1 );
+	const [ resizeListener, { width: totalWidth } ] = useResizeObserver();
+
+	const referenceCell =
+		gridRef?.current?.querySelector( '[role="gridcell"]' ) || {};
```

Comment:
Instead of querying the DOM, would it make sense to add a `ref` in `Grid` component and then do something like: `getComputedStyle(gridRef.current)?.getPropertyValue('grid-template-columns')`? Then we could derive the column count.